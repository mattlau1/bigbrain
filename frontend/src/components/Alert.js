import React, { useState } from 'react';
import PropTypes from 'prop-types';

const Alert = props => {
  const [exit, setExit] = useState(false);
  const [progress, setProgress] = useState(0);
  const [intervalID, setIntervalID] = useState(null);
  const tickSpeed = 20
  const maxTime = 400

  // starts the timer for alert
  const handleStartTimer = () => {
    const id = setInterval(() => {
      setProgress(prev => {
        if (prev < 100) {
          return prev + 0.5;
        }

        clearInterval(id);
        return prev;
      });
    }, tickSpeed);
    setIntervalID(id);
  };

  // closing the alert
  const handleCloseAlert = () => {
    clearInterval(intervalID);
    setExit(true);
    setTimeout(() => {
      props.dispatch({
        type: 'REMOVE_ALERT',
        id: props.id
      })
    }, maxTime)
  };

  // close if progress of progress bar is 100
  React.useEffect(() => {
    if (progress === 100) {
      handleCloseAlert()
    }
  }, [progress])

  // start timer when alert appears
  React.useEffect(() => {
    handleStartTimer();
  }, []);

  return (
    <div
      className={`alert-item d-none d-md-block d-lg-block d-xl-block ${
        props.type === 'SUCCESS' ? 'success' : 'error'
      } ${exit ? 'exit' : ''}`}
    >
      <p className="m-0 p-2">{props.message}</p>
      <div className={'bar'} style={{ width: `${progress}%` }} />
    </div>
  );
};

Alert.propTypes = {
  type: PropTypes.string,
  message: PropTypes.string,
  dispatch: PropTypes.func,
  id: PropTypes.string
};

export default Alert;
