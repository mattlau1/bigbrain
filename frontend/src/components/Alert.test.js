import { shallow } from 'enzyme';
import React from 'react';
import Alert from './Alert.js';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';

describe('Alert', () => {
  it('should show a successful alert with the right message', () => {
    const tree = renderer.create(<Router><Alert /></Router>).toJSON();
    const wrapper = shallow(<Alert type='SUCCESS' message='This is a success'/>);
    const text = wrapper.find('div p');
    expect(text.text()).toBe('This is a success');
    expect(tree).toMatchSnapshot();
  })

  it('should show an unsuccessful alert with the right message', () => {
    const tree = renderer.create(<Router><Alert /></Router>).toJSON();
    const wrapper = shallow(<Alert type='ERROR' message='This is an error'/>);
    const text = wrapper.find('div p');
    expect(text.text()).toBe('This is an error');
    expect(tree).toMatchSnapshot();
  })

  it('should show a successful alert', () => {
    const tree = renderer.create(<Router><Alert /></Router>).toJSON();
    const wrapper = shallow(<Alert type='SUCCESS' message='Cat says meow'/>);
    expect(wrapper.hasClass('success')).toBe(true);
    expect(wrapper.hasClass('error')).toBe(false);
    expect(tree).toMatchSnapshot();
  })

  it('should show an unsuccessful alert', () => {
    const tree = renderer.create(<Router><Alert /></Router>).toJSON();
    const wrapper = shallow(<Alert type='ERROR' message='Dog says woof'/>);
    expect(wrapper.hasClass('success')).toBe(false);
    expect(wrapper.hasClass('error')).toBe(true);
    expect(tree).toMatchSnapshot();
  })

  it('should try to exit after 4 seconds', () => {
    const wrapper = shallow(<Alert type='SUCCESS' message='Hello world'/>);

    // alert should apply a class 'exit' to itself and remove itself from
    // the dom after 4 seconds have passed
    setTimeout(() => {
      expect(wrapper.hasClass('exit')).toEqual(true);
    }, 4000);
  })

  it('should be closed after 4 seconds', () => {
    const wrapper = shallow(<Alert type='SUCCESS' message='Hello'/>);
    setTimeout(() => {
      expect(wrapper.exists()).toBeFalsy();
    }, 5001);
  })
});
