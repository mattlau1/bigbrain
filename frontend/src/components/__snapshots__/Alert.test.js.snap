// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should show a successful alert 1`] = `
<div
  className="alert-item d-none d-md-block d-lg-block d-xl-block error "
>
  <p
    className="m-0 p-2"
  />
  <div
    className="bar"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`Alert should show a successful alert with the right message 1`] = `
<div
  className="alert-item d-none d-md-block d-lg-block d-xl-block error "
>
  <p
    className="m-0 p-2"
  />
  <div
    className="bar"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`Alert should show an unsuccessful alert 1`] = `
<div
  className="alert-item d-none d-md-block d-lg-block d-xl-block error "
>
  <p
    className="m-0 p-2"
  />
  <div
    className="bar"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`Alert should show an unsuccessful alert with the right message 1`] = `
<div
  className="alert-item d-none d-md-block d-lg-block d-xl-block error "
>
  <p
    className="m-0 p-2"
  />
  <div
    className="bar"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;
